### ISSUE TITLE
API access to open file's original encoding

### ISSUE THREAD

**Issue by Colengms** [microsoft/vscode#209503]
This is a request for API access to the original encoding of a file.  i.e.:

![image](https://github.com/microsoft/vscode/assets/49173979/59620a1b-4a31-4ddb-8a8b-1c096375e9f3)

(If this is already possible, please let me know. I couldn't find it.)

If I understand correctly, VS Code represents the content of open files to extensions as UTF-8.  In the C/C++ Extension, there are scenarios in which IntelliSense should reflect how the compiler would interpret the file, which may depend on the files actual encoding. i.e. MSVC and string literal contents.

**Comment by Colengms** [comment]
In addition to the original encoding, it would also be very helpful to have access to (a bool indicating) whether or not the file was re-encoded from the encoding specified by the `files.encoding` setting (meaning that the file encoding was not able to be auto-detected).  The C/C++ Extension indexed files in the workspace, honoring the `files.encoding` setting when reading directly from disk (if unable to detect the encoding).  When that setting changes we should reparse files, preferably only the ones impacted by that change. We can determine ourselves for files read directly from disk.  However, if open in the editor, we use the file buffer provided by VS Code. A change to `files.encoding` could indicate that the buffer we had used (from VS Code) may not have been translated correctly and should be reparsed.

**Comment by bpasero** [comment]
Related: https://github.com/microsoft/vscode/issues/824

**Comment by bpasero** [comment]
I would like to go ahead and merge into https://github.com/microsoft/vscode/issues/824 as per the original request. https://github.com/microsoft/vscode/issues/209503#issuecomment-2537626506 seems unrelated, so I would suggest to report as individual new issue.

**Comment by bpasero** [comment]
@Colengms 👋 friendly ping for request for input: https://github.com/microsoft/vscode/issues/824#issuecomment-2657082554

**Comment by Colengms** [comment]
Hi @bpasero .  Re: https://github.com/microsoft/vscode/issues/824#issuecomment-2657082554 , would it be possible to throw in just an additional boolean to indicate whether the encoding provided was detected as a result of a guess, due to `autoGuessEncoding` being set to true?  The C/C++ Extension loads files directly from disk (such as headers that are not open in the editor but #include'd by an open file) from a native process.  If that header is then opened in the editor, we switch to using that buffer, and can encounter some offset inconsistencies internally, if our interpretation of the file doesn't match VS Code's guess. That bool could help us detect and correct that.

**Comment by bpasero** [comment]
I do not think we would add that to the API given the very little use of that `autoGuessEncoding` encoding setting through our user base. Why would you need this specifically?

**Comment by Colengms** [comment]
> Why would you need this specifically?

Hi @bpasero .  If, in additional to the original encoding, we could get this boolean, and it were false, it would indicate that no further work is needed.  It's only when `files.autoGuessEncoding` is used that we have further work to do.  We're already tracking changes to `files.encoding` and interpreting non-guessed encoding the same as VS Code.  It's relevant, as it disambiguates whether or not the encoding reported is the actual encoding of the file (based on BOM, common UTF-16 detection, or explicit user preference) vs. a guess.  It would seem pretty simple to include.

Note that there are some scenarios our users are running into that may currently only be addressable using `files.autoGuessEncoding`, as `files.encoding` is too heavy a hammer, impacting both user's files, their system includes, and all of their library headers.  https://github.com/microsoft/vscode-cpptools/issues/11407#issuecomment-1714398813

**Comment by bpasero** [comment]
What specifically is the difference between a text document where the encoding was guessed vs. when the encoding was configured by the user or set explicitly after opening?

**Comment by bpasero** [comment]
@Colengms btw latest VS Code insiders ships with the `encoding` property, maybe you can already use it?

https://github.com/microsoft/vscode/blob/80b20f33e756ec02365fd91d39a4ba20c5b39e7e/src/vscode-dts/vscode.proposed.textDocumentEncoding.d.ts#L6-L30

**Comment by heartacker** [comment]
thanks

**Comment by Colengms** [comment]
> btw latest VS Code insiders ships with the encoding property, maybe you can already use it?

Hi @bpasero .  Yes. We're now using this, and it's very helpful.  Thanks.  :)
